lib/manager/components/transform/FeedTransformation.js

```diff
@@ -23,124 +25,56 @@ type Props = {
-          ? transformation.csvData
-          : this.state.csvData
-        const textValue = csvData || ''
-        const numLines = !textValue ? 0 : textValue.split(/\r*\n/).length
-        fields.push(
-          <div>
-            <label htmlFor='csvData'>
-              Add the CSV data to add to/replace in the incoming GTFS:
-              <textarea
-                id='csvData'
-                onChange={this._onChangeCsvData}
-                style={{
-                  fontFamily: 'monospace',
-                  fontSize: 'x-small',
-                  height: '80px',
-                  overflow: 'auto',
-                  width: '400px',
-                  whiteSpace: 'pre'
-                }}
-                placeholder={
-                  `stop_id,stop_code,stop_name,stop_lat,stop_lon
-  1234567,188390987,Broad Ave,33.98768,-87.72686`
-                }
-                value={textValue} />
-            </label>
-            <div style={{marginBottom: '10px'}}>
-              <Button
-                bsSize='xsmall'
-                disabled={inputIsUnchanged}
-                style={{marginRight: '5px'}}
-                onClick={this._onSaveCsvData}>
-                Save CSV
-              </Button>
-              <small>{numLines} lines</small>
-            </div>
-          </div>
-        )
-        break
-      default:
-        break
-    }
-    return fields
+    const {feedSource, index, onChange, transformation} = this.props
+    const FieldsComponent = transformationTypes[type].component
+    return FieldsComponent && (
+      <FieldsComponent
+        feedSource={feedSource}
+        index={index}
+        onSave={onChange}
+        transformation={transformation}
+      />
+    )
   }
 
   _getValidationIssues = () => {
     const {feedSource, transformation} = this.props
-    // Get selected version (if applicable).
-    const version = feedSource.feedVersions &&
-      feedSource.feedVersions.find(v => v.id === transformation.sourceVersionId)
-    const issues = []
+    let issues: Array<string> = []
     if (!transformation.table) {
       issues.push('Table must be defined')
     }
-    switch (transformation['@type']) {
-      case 'ReplaceFileFromVersionTransformation':
-        // Only trigger validation issue if versions have been loaded.
-        if (!version && feedSource.feedVersions) {
-          issues.push('Version must be defined')
-        }
-        break
-      case 'ReplaceFileFromStringTransformation':
-        if (!transformation.csvData) {
-          issues.push('CSV Data must be defined')
-        }
-        // TODO: Further validate CSV data based on parsing fields, checking for
-        //   missing columns, etc.
-        break
-      default:
-        break
+
+    const {getValidationIssues} = transformationTypes[transformation['@type']]
+    if (typeof getValidationIssues === 'function') {
+      issues = issues.concat(getValidationIssues(transformation, feedSource))
     }
```

Comment:
This sends over the `transformation` prop and not the values in the current state of the transformation. So if the user edits the NormalizeFieldTransformation for example, the validation issues won't go away until the transformation is persisted.

Reply:
Thanks for pushing this. I moved validation code of the editing state inside each transform component in https://github.com/ibi-group/datatools-ui/pull/649/commits/ccc561232dfc744f574b02b6b35c44cdf0a73f9b.